Feature: new assets REG, REG Vote Power,  RWA realtime price, Settings updates

BenoistP opened the pull request:
- Prices for RWA & REG are fetched from Honeyswap (UniV2) pools (average of 2 pools: asset/USDC and asset/wXDAI)
- Filter 'other' assets (i.e. others than RealTokens) in views using a switch (Others assets) in settings
- domain static.debank.com is used for diplaying REG image
next.config.js
- small fix when address has no corresponding user ('User not found' error)
src/components/assetsView/AssetsViewSearch.tsx
- Misc updates in Setting menu
  - icons in switches
- Labels optional icon
- Interfaces RWARealtoken, REGRealtoken, REGVotingPowertoken are the same thing as OtherRealtoken (previously named RWARealtoken)
- ABIs
  - ERC20: Full abi
  - UniswapV2FactoryABI
- Consts: tokens & factory addresses, decimals, default prices
consts/otherTokens.ts

BenoistP committed:
first version : Assets as rows

BenoistP committed:
'other' assets display switch; 'other' assets shown in Summary

'other' assets = tokens others than RealTokens = RWA, REG, ...

- Display switch for hiding 'other' assets in table rows
- 'other' assets figures added to Summary section

BenoistP committed:
contracts utils

BenoistP committed:
use contract utils for fetching balances

BenoistP committed:
Eth provider removed: REG Voting Power is not deployed on Eth

BenoistP committed:
batchCallOneFunction

not fully tested

BenoistP committed:
error args

BenoistP committed:
comments

BenoistP committed:
wrong 'inheritance'

BenoistP committed:
smol fix: properly throw 'User not found' error

BenoistP committed:
style/comments

BenoistP committed:
settings labels and icons

BenoistP committed:
unitIcon: optional + alignment with other units

BenoistP committed:
Filter 'other' assets in Grid view

BenoistP committed:
Settings: switch icons, section names

BenoistP committed:
fix: lint warning

BenoistP committed:
fix: lint warning

BenoistP committed:
fix: build error ?

BenoistP committed:
fix: build error #2 ?

BenoistP committed:
fix: build error #3 ?

BenoistP committed:
fix: build error #4 ?

BenoistP committed:
fix: build error #5 ?

BenoistP committed:
Fix: existing filter error when filtering on "Owned on"

jycssu-com committed:
fix: avoid fetching others assets balances if user wallet was not set

jycssu-com commented:
Everythings seems good to me, i just added a check for fetching other assets balances only if the addressList is not empty for avoid a double request on app initialization

Despite this, there are still a large number of RPC requests going out, but it seems to be related to the RPC initialization mechanism to look for a working one, so out of the scope of this PR (maybe we need another PR to improve this part)

@NandyBa I'll let you review it on your side too, and approve if everything is good for you (I can't since I committed)

jycssu-com committed:
feat: add other token price on cards and fix decimals rounds for total invest

jycssu-com committed:
fix: improve subsudy filter for exclude other realtokens

jycssu-com committed:
fix: fix last changes filter and use correct typing in filters for avoid futurs errors

jycssu-com committed:
feat: set initial launch date for RWA token

jycssu-com committed:
feat: initialize rpc provider only once

jycssu-com committed:
feat: include reg tokens locked inside the incentive vault

jycssu-com committed:
fix: fix eslint warnings and add missing translations

NandyBa commented on src/hooks/useREG.ts:
```diff
@@ -0,0 +1,117 @@
+  const totalTokens = Number(contractRegTotalSupply) / 10 ** REGtokenDecimals
+  const amount = totalAmount / 10 ** REGtokenDecimals
+
+  const regPriceUsdc = await getUniV2AssetPrice(
+    HoneySwapFactory_Address,
+    REG_ContractAddress,
+    USDConXdai_ContractAddress,
+    REGtokenDecimals,
+    USDCtokenDecimals,
+    GnosisRpcProvider,
+  )
+  const regPriceWxdai = await getUniV2AssetPrice(
+    HoneySwapFactory_Address,
+    REG_ContractAddress,
+    WXDAI_ContractAddress,
+    REGtokenDecimals,
+    WXDAItokenDecimals,
+    GnosisRpcProvider,
+  )
+
+  const averagePrice = averageValues([regPriceUsdc, regPriceWxdai])
```
Please use `xDAIUSD` and `USDCUSD` price. You can found an example in `PropertityCard`

BenoistP replied:
Is there a source for USDC price ?
I was looking at
[/home/dev/git/RealToken-Community/dashboard-v2/src/store/features/currencies/currenciesSelector.ts](https://github.com/RealToken-Community/dashboard-v2/blob/d23c1d46f97211ae5bfcb28bbf68973d203f31f0/src/store/features/currencies/currenciesSelector.ts#L4)

https://github.com/RealToken-Community/dashboard-v2/blob/d23c1d46f97211ae5bfcb28bbf68973d203f31f0/src/store/features/currencies/currenciesSlice.ts#L4

but couldn't find one, only XDAI price

BenoistP replied:
still can't find a source for USDC price 🤔

BenoistP replied:
Still no answer to my questions.
There's nothing related t prices to be found in src/components/cards/main/PropertiesCard.tsx as far as I can understand.

NandyBa replied:
Sorry I miss, I was thinking that the were `xdaiusd` and `usdcusd` but there is only xdaiusd in src/repositories/currencies.repository.ts

NandyBa replied:
Address for usdcusd pricess feed is 0x26C31ac71010aF62E6B486D1132E266D6298857D and come from https://data.chain.link/feeds/xdai/mainnet/usdc-usd

BenoistP replied:
Ok, thanks, I'll check it

NandyBa commented on src/utils/blockchain/contract.ts:
```diff
@@ -0,0 +1,71 @@
+    do {
+      // wait if attempt > 0 and grow wait time for each attempt
+      attempt && wait(BATCH_WAIT_BETWEEN_ATTEMPTS_MS * attempt)
+      attempt++
+      try {
+        let results: object[] = []
+        // Split the array into chunks
+        const chunks = []
+        // Divide chunk size by attempt at each iteration (decrease chunk size for each attempt)
+        const chunkSize = MAX_BATCH_CONTRACT_PER_CHUNK / attempt
+        for (let i = 0; i < _argsArray.length; i += chunkSize) {
+          chunks.push(_argsArray.slice(i, i + chunkSize))
+        }
+        for (let i = 0; i < chunks.length; i++) {
+          const chunkPromises: object[] = []
+          const _argsChunk = chunks[i]
+          _argsChunk.forEach(async (_args /* , idx */) => {
+            chunkPromises.push(contractCall(_contract, _methodName, _args))
+          })
+          const chunkResults = await Promise.all(chunkPromises)
+          results = results.concat(chunkResults)
+          // wait between remaining chunks
+          if (i < chunks.length - 1) {
+            wait(BATCH_WAIT_BETWEEN_CHUNKS_MS)
+          }
+        }
+        return results
+      } catch (error) {
+        const chainId =
+          (await _contract?.runner?.provider?.getNetwork())?.chainId ??
+          'unknown'
+        console.error(
+          `batchCallOneContractOneFunctionMultipleParams Error:: chainId: ${chainId} contract address: ${_contract?.target} methodName: ${_methodName} args: [${_argsArray}]`,
+        )
+      }
+    } while (attempt < BATCH_MAX_ATTEMPTS)
+  } catch (error) {
+    console.error(error)
+  }
+}
```
Maybe better to define a batch number per query with a default value.
Idea to reduce batch size per attend is good

BenoistP replied:
What do you mean by "batch number" ?
Batch query is starting with a size of 100 queries per batch call (second call is 100/2 = 50, third 100/3=33, ...), 
`const MAX_BATCH_CONTRACT_PER_CHUNK = 100`
If there is 555 calls to process it'll split like this
- first attempt: 100, 100, 100, 100, 100, 55
- second attempt: 50, 50, 50, 50, 50, 50, 50, 50, 50, 50,  50, 5
- third: 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 27
- 4: 25 (x 22) , 5
- 5: 20 (x 27) , 15

NandyBa replied:
I mean a batch size per query

NandyBa replied:
function1 -> batch size 100
function2 -> batch size 125
function3 -> batch size undefined -> 100
function4 -> batch size 30

BenoistP replied:
just pushed batch size parameters for customs calls 🫡

BenoistP committed:
Update .gitignore

exclude pnpm lockfile

BenoistP committed:
Create .nvmrc

add .nvmrc for switching between node versions

BenoistP committed:
Update contract.ts

optional  parameters for customizing batch calls: batch max and min size, number of attempts before giving up

BenoistP committed:
Regvotingpower: icon orange, size grows depending on amount

BenoistP committed:
unused import

BenoistP committed:
Usdc currency/rate

BenoistP committed:
Voting power size & fill color depending on power amount

BenoistP committed:
Voting power size & fill color depending on power amount

BenoistP committed:
Fixes: token prices in USD, avoid fetching REG vault balances on Eth

- Tokens values fix: Updated assets prices by using currenciesRates and user currency rate (previously considering usdc = usd and xdai = usd as well),
now assets prices are converted from their respective currencies (e.g. usdc, xdai) to usd, then total value is priced in user choosen currency.
- Smol fix: getAddressesLockedBalances was fetching vault balances on Eth, throwing errors as there is not such vault on Eth.
Added parameters for fetching any number of vault by provider(s).

BenoistP committed:
comment

BenoistP committed:
Check Providers ability to handle requests

- Check Providers ability to handle requests
- Additionnal providers
https://rpc.ankr.com/eth and https://eth-pokt.nodies.app currently fail to handle concurrent requests

BenoistP commented:
- I've added a test for checking providers ability to handle concurrent requests:
(On ETH https://rpc.ankr.com/eth and https://eth-pokt.nodies.app currently fail to resolve balances when "too much" requests are sent)
- Smol fix: getAddressesLockedBalances was fetching vault balances on Eth
- Added 2 additionnal providers for each chain
- Assets prices are now converted from USDC/XDAI to USD before being displayed in user choosen currency
- Additionnal parameters on batchCallOneContractOneFunctionMultipleParams for fine tuning calls (batch sizes, attempts)

BenoistP committed:
Assets hooks moved in index for avoiding duplicate requests

- Both SummaryCard and AssetsView use the same hooks for getting their data, leading to duplicate web3 requests
Moving hooks in parent components prevents this form happening.

BenoistP commented:
Hooks calls are duplicated in 2 components (SummaryCard and AssetsView) so the same requests are made when fetching data.
Moving it to index solves it, let me know if it's considered "dirty" i'll revert the last commit.

BenoistP committed:
prettier

NandyBa commented on src/components/assetsView/filters/AssetsViewSort.tsx:
```diff
@@ -92,51 +92,58 @@ export const AssetsViewSort: FC<AssetsViewSortProps> = ({
       case AssetSortType.OCCUPANCY:
-        return B.rentedUnits / B.totalUnits - A.rentedUnits / A.totalUnits
+        return (
+          (b.rentedUnits ?? 0) / (b.totalUnits ?? 0) -
+          (a.rentedUnits ?? 0) / (a.totalUnits ?? 0)
```
Warning potential divison by zero

BenoistP replied:
Fix: in case asset.totalUnits is unknown/zero (should not happen)
Divide by ONE instead of ZERO

NandyBa commented on src/utils/blockchain/regVault.ts:
```diff
@@ -0,0 +1,127 @@
+                console.warn('ABI, contract address or function name missing')
+              return null
+            }
+            const RegVaultGetUserGlobalStateContract = new Contract(
+              contractAddress,
+              abi,
+              provider,
+            )
+            const state = batchCallOneContractOneFunctionMultipleParams(
+              RegVaultGetUserGlobalStateContract,
+              functionName,
+              addressListObj,
+            )
+            return state
+          },
+        )
+      },
+    )
+
+    // Wait all promises to resolve and flatten the array
+    const statesArray = await Promise.all(statesPromises.flat())
+    const states = statesArray.flat()
+    // Sum all valid balances
+    states.forEach((state: object | null | undefined) => {
+      try {
+        if (state) {
+          totalAmount += Number((state as { 0: string; 1: string })['0'])
+        }
+      } catch (error) {
+        if (consoleWarnOnError) {
+          console.warn('Failed to sum balances', error)
+        }
+      }
+    })
+    return totalAmount
+  } catch (error) {
+    console.warn('Failed to get balances', error)
+  }
+  return totalAmount
+}
```
I think it can be improve but let do that on an other PR

BenoistP replied:
Thanks, I've planned to enhance things in coming weeks

BenoistP committed:
Avoid divide by zero in case asset.totalUnits is unknown/zero

NandyBa reviewed (commented):
When loading the dashboard locally I got 18 properties instead of 20. And when I try a search of RWA on the search bar I don't have it.

BenoistP committed:
"Clean" scripts

BenoistP committed:
Assets filtering moved to the right place

filter has to be in assetsView/AssetsView.tsx

NandyBa reviewed (commented):
Still present
<img width="1507" alt="Capture d’écran 2024-12-13 à 19 50 12" src="https://github.com/user-attachments/assets/b436b61e-799e-4c76-9d62-361d3c8fd87d" />


Normal comportment: <img width="1507" alt="Capture d’écran 2024-12-13 à 20 01 51" src="https://github.com/user-attachments/assets/2b8d6dbd-fe9f-4dc9-96c3-41e38f4b6990" />


Config: no option enabled

BenoistP committed:
Paging translation labels: placeholder, All

BenoistP committed:
simplify pageSize handling

BenoistP committed:
Assets filtering fix (causing pagination issues)

OtherAssets filtering was improperly implemented into the map loops by skipping non "others assets" value, causing holes in display (missing cards).
Filtering is now done in AssetsView where it should have been in the first place

BenoistP commented:
Should be fixed now, translated paging labels at the same time

NandyBa commented on src/i18next/locales/en/common.json:
```diff
@@ -19,16 +19,20 @@
-    "realtime": "Realtime",
-    "global": "Global",
+    "rents": "Rents calculation",
+    "realtime": "Real",
+    "future": "Future",
```
Please to not change. Do not really correspond to the purpose of the mode

BenoistP replied:
right, done

NandyBa commented on src/repositories/RpcProvider.ts:
```diff
@@ -1,47 +1,215 @@
+        provider,
+        REG_ContractAddress,
+        5,
+        5,
+        5,
+        150,
+      )
+      if (rpcThresholdValue < 1) {
+        // Throw error if the threshold is 0
+        // Means the provider is not able to handle required concurrent requests number
+        // skip it and try next one
+        throw new Error('rpcThresholdValue returned 0')
+      }
+      // If any error has occurred before, log the successful connection
+      if (failedRpcErrorCount > 0) {
+        console.info(
+          `Successfully connected to ${url} after ${failedRpcErrorCount} failed attempts`,
+        )
+      }
+      return provider
     } catch (error) {
-      console.error(`Failed to connect to ${url}, trying next one...`)
+      failedRpcErrorCount++
+      if (!rpcConnectOk) {
+        // Connection error
+        console.error(`Failed to connect to ${url}, trying next one...`, error)
+      } else if (rpcThresholdValue < 1) {
+        // Threshold error
+        console.error(
+          `Successfull connection to ${url} BUT failed to test rpcThresholdValue, trying next one...`,
+          error,
+        )
+      } else {
+        // General error
+        console.error(`Failed to connect to ${url}, trying next one...`, error)
+      }
     }
   }
+  throw new Error(`All RPC URLs (${urls?.length}) failed`)
+}
```
This part slow down the page loading

NandyBa replied:
I recommend to run it appart from the front

BenoistP replied:
I'll see if i can improve things but if these tests aren't run then there's no garantee next batchs requests will succeed and nothing will be loaded if the rpc can't support multiple requests at the same time (like https://rpc.ankr.com/eth seems like not supporting it)

NandyBa replied:
Can we populate the code with test results and run tests only for new RPCs, instead of running all tests every time? I don’t want to block the PR, but I think merging it like this is manageable in the develop branch. However, it might create blockers later when escalating to preprod and prod.

BenoistP replied:
Checks for testing rpc ability to handle batch requests only run once for each rpc, one after the other like it is already the case.
Currently rpc connection checks take around 700ms, additionnal batch check requires 100-150 ms more.
If the first rpcs fails to respond and to handle batch request, the second is tested and so on.

BenoistP committed:
Revert changes: realtime/global labels for rents calculation switch

Sigri44 commented:
@NandyBa Where are we now? Are there any requests that don't concern the original PR, if I understand correctly?

NandyBa commented:
> @NandyBa Where are we now? Are there any requests that don't concern the original PR, if I understand correctly?
???
Currently all is good expect: https://github.com/RealToken-Community/dashboard-v2/pull/91#discussion_r1885109100

NandyBa reviewed (approved):
Approved. We will revisit the RPC test if we find it annoying
